cmd/node_problem_detector.go

```diff
@@ -64,7 +56,8 @@ func getNodeNameOrDie() string {
 	// 2. For some cloud providers, os.Hostname is different from the real hostname.
 	nodeName = os.Getenv("NODE_NAME")
 	if nodeName != "" {
-		return nodeName
+		npdo.HostnameOverride = nodeName
```

Comment:
* Add another field `NodeName` in `NodeProblemDetectorOptions`. With this new field, we will not change `HostnameOverride` field anymore. :) However, we still need to change `NodeName` field which is the actual node name used to communicate with Kubernetes Apiserver. IMO, this should be fine, we have preserve all the command line arguments unchanged. WDYT @Random-Liu .
* Move `setNodeNameOrDie` to be a method for `*NodeProblemDetectorOptions`.